streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java

```diff
@@ -443,14 +449,84 @@ public void shouldReturnThreadMetadata() {
     }
 
     @Test
-    public void testCleanup() {
+    public void shouldAllowCleanupIfApplicationIsDown() throws Exception {
         final StreamsBuilder builder = new StreamsBuilder();
         final KafkaStreams streams = new KafkaStreams(builder.build(), props);
 
+        final File taskDir = new File(
+            props.getProperty(StreamsConfig.STATE_DIR_CONFIG)
+                + File.separator + props.getProperty(StreamsConfig.APPLICATION_ID_CONFIG)
+                + File.separator + new TaskId(0, 0));
+        final File testFile = new File(taskDir, "test-file");
+
+        assertTrue(taskDir.mkdirs());
+        assertTrue(testFile.createNewFile());
+
         streams.cleanUp();
+        assertFalse(testFile.exists());
+        assertFalse(taskDir.exists());
+
         streams.start();
         streams.close();
+
+        assertTrue(taskDir.mkdirs());
+        assertTrue(testFile.createNewFile());
+
         streams.cleanUp();
+        assertFalse(testFile.exists());
+        assertFalse(taskDir.exists());
+    }
+
+    @Test
+    public void shouldFailWithOverlappingFileLockException() throws Exception {
```

Comment:
Could we add two more positive tests and see if we would reproduce the issue reported in KAFKA-6647?

1) in thread A: create the file, open the channel, grab the lock, and then release the lock, and close the channel; and then call `cleanup` which should do the above again, and we should not expect any exception.

1) in thread A: create the file, open the channel, grab the lock, and then release the lock, and close the channel; and then in a different thread B, call `cleanup` which should do the above again, and we should not expect any exception.